tokio/src/coop.rs

```diff
@@ -29,384 +31,188 @@
+//! other futures. By doing this, you avoid double-counting each iteration of
+//! the outer future against the cooperating budget.
 //!
-//!   [`poll`]: https://doc.rust-lang.org/std/future/trait.Future.html#tymethod.poll
+//! [`poll`]: https://doc.rust-lang.org/std/future/trait.Future.html#tymethod.poll
 
 // NOTE: The doctests in this module are ignored since the whole module is (currently) private.
 
 use std::cell::Cell;
-use std::future::Future;
-use std::pin::Pin;
 use std::task::{Context, Poll};
 
-/// Constant used to determine how much "work" a task is allowed to do without yielding.
-///
-/// The value itself is chosen somewhat arbitrarily. It needs to be high enough to amortize wakeup
-/// and scheduling costs, but low enough that we do not starve other tasks for too long. The value
-/// also needs to be high enough that particularly deep tasks are able to do at least some useful
-/// work at all.
-///
-/// Note that as more yield points are added in the ecosystem, this value will probably also have
-/// to be raised.
-const BUDGET: usize = 128;
+thread_local! {
+    static CURRENT: Cell<Budget> = Cell::new(Budget::unconstrained());
+}
 
-/// Constant used to determine if budgeting has been disabled.
-const UNCONSTRAINED: usize = usize::max_value();
+/// Opaque type tracking the amount of "work" a task may still do before
+/// yielding back to the scheduler.
+#[derive(Debug, Copy, Clone)]
+pub(crate) struct Budget(Option<u8>);
 
-thread_local! {
-    static HITS: Cell<usize> = Cell::new(UNCONSTRAINED);
+impl Budget {
+    /// Budget assigned to a task on each poll.
+    ///
+    /// The value itself is chosen somewhat arbitrarily. It needs to be high
+    /// enough to amortize wakeup and scheduling costs, but low enough that we
+    /// do not starve other tasks for too long. The value also needs to be high
+    /// enough that particularly deep tasks are able to do at least some useful
+    /// work at all.
+    ///
+    /// Note that as more yield points are added in the ecosystem, this value
+    /// will probably also have to be raised.
+    const fn initial() -> Budget {
+        Budget(Some(128))
+    }
+
+    /// Returns an unconstrained budget. Operations will not be limited.
+    const fn unconstrained() -> Budget {
+        Budget(None)
+    }
+
+    fn has_remaining(&self) -> bool {
+        self.0.map(|budget| budget > 0).unwrap_or(true)
+    }
+
+    /// Decrement the budget. Returns `true` if successful. Decrementing fails
+    /// when there is not enough remaining budget.
+    fn decrement(&mut self) -> bool {
+        if let Some(num) = &mut self.0 {
+            if *num > 0 {
+                *num -= 1;
+                true
+            } else {
+                false
+            }
+        } else {
+            true
+        }
+    }
 }
 
-/// Run the given closure with a cooperative task budget.
-///
-/// Enabling budgeting when it is already enabled is a no-op.
```

Comment:
Was this behavior being used somewhere previously?

Reply:
no tests as far as I can see. Based on LocalSet, it seems like it is the desired behavior.